python/tchannel/tornado/peer.py

```diff
@@ -188,18 +205,23 @@ def choose(self, hostport=None, score_threshold=None):
             If specified, Peers with a score equal to or below this will be
             ignored. Defaults to the value specified when the PeerGroup was
             initialized.
+        :param blacklist:
+            Peers on the blacklist won't be chosen.
         :returns:
             A Peer that matches all the requested criteria or None if no such
             Peer was found.
         """
-        if hostport:
+
+        blacklist = blacklist or set()
+        if hostport and hostport not in blacklist:
             return self.get(hostport)
 
         score_threshold = score_threshold or self._score_threshold or 0
         chosen_peer = None
         chosen_score = 0
 
-        for host in self.hosts:
+        hosts = set(self.hosts) - blacklist
```

Comment:
[dict.viewkeys](https://docs.python.org/2/library/stdtypes.html#dict.viewkeys) give you a set-like object that contains the keys of the dictionary. Consider using that.